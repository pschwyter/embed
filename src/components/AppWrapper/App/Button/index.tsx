import classnames from "classnames";
import Client from "models/Client";
import DialogueSvg from "icons/Dialogue.svg";
import SVG from "react-inlinesvg";
import { Component, h } from "preact";
import "./style.scss";

const NOTIFICATION_BADGE_SIZE_RATIO = 0.386;
const NOTIFICATION_BADGE_BOTTOM_RATIO = 0.705;
const NOTIFICATION_BADGE_RIGHT_RATIO = -0.09;
const NOTIFICATION_BADGE_BORDER_RATIO = 0.352;

interface InterfaceButton {
  client: Client,
  showIntroEmoji: boolean,
  showNotification: boolean,
  isDraggable: boolean,
  toggleChat(): void
}

export default class Button extends Component<InterfaceButton> {
  state = {
    customIconHasLoaded: false
  };

  setCustomIconHasLoaded = () => {
    this.setState({ customIconHasLoaded: true });
  }

  get notificationStyles() {
    const { client } = this.props;
    const buttonSize = client.chatButton.size;
    const heightAndWidth = Math.ceil(buttonSize * NOTIFICATION_BADGE_SIZE_RATIO);

    return {
      height: heightAndWidth,
      width: heightAndWidth,
      bottom: buttonSize * NOTIFICATION_BADGE_BOTTOM_RATIO,
      right: buttonSize * NOTIFICATION_BADGE_RIGHT_RATIO,
      borderWidth: heightAndWidth * NOTIFICATION_BADGE_BORDER_RATIO
    };
  }

  render(props: InterfaceButton) {
    const { customIconHasLoaded } = this.state;
    const {
      toggleChat, client, showIntroEmoji,
      showNotification, isDraggable
    } = props;

    const useCustomIcon = client.chatButton.icon_type === "custom" || null;
    const chatButtonURL = useCustomIcon ? client.chatButton.icon_path : "static/icons/Dialogue.svg";
    const buttonSize = client.chatButton.size;

    const adaEmbedButtonContainerClassName = classnames("ada-embed-button-container", {
      "ada-embed-button-container--loading": useCustomIcon && !customIconHasLoaded,
      "ada-embed-button-container--not-draggable": !isDraggable
    });
    const chatButtonIconClassname = classnames("ada-embed-button__icon", {
      "ada-embed-button__icon--hide": showIntroEmoji
    });

    return (
      <div className={adaEmbedButtonContainerClassName}>
        <button
          title="Open Support Chat"
          accessKey="1"
          className="ada-embed-button"
          onClick={toggleChat}
          style={{
            width: buttonSize,
            height: buttonSize,
            backgroundColor: client.chatButton.background_color
          }}
        >
          { useCustomIcon ?
            <SVG
              src={chatButtonURL}
              className={chatButtonIconClassname}
              onLoad={this.setCustomIconHasLoaded}
              cacheGetRequests={true}
            /> :
            <DialogueSvg
              className={chatButtonIconClassname}
            />
          }
          {showIntroEmoji && (
            <img
              alt=""
              role="presentation"
              src={client.intro.body}
              className={
                classnames("ada-embed-button__emoji", {
                  "ada-embed-button__emoji--show": showIntroEmoji
                })
              }
            />
          )}
        </button>
        {showNotification && (
          <div
            className="ada-embed-notification"
            alt="New Message"
            role="alert"
            style={this.notificationStyles}
          />
        )}
      </div>
    );
  }
}
