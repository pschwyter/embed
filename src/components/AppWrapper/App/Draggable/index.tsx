import { Component, h, cloneElement } from "preact";
import { outlineSuppressionHandlers } from "services/a11y";
import { isIOS } from "services/browsers";

interface DraggableProps {
  x?: number,
  y?: number,
  updatePosition?: (x: number, y: number) => void
}

interface DraggableState {
  dragging: boolean,
  rel: { x: number, y: number },
  pos: { x: number, y: number },
  beingDragged: boolean,
  wasMoved: boolean
}

interface TouchEventItem {
  pageX: number,
  pageY: number
}

export default class Draggable extends Component<DraggableProps, DraggableState> {
  static defaultProps = {
    x: 0,
    y: 0
  };
  myRef: any;

  constructor(props: DraggableProps) {
    super(props);

    const { x, y } = props;

    this.state = {
      pos: { x, y },
      dragging: false,
      rel: null,
      beingDragged: false,
      wasMoved: false
    };

    this.onMouseMove = this.onMouseMove.bind(this);
    this.onMouseDown = this.onMouseDown.bind(this);
    this.onTouchStart = this.onTouchStart.bind(this);
    this.onTouchMove = this.onTouchMove.bind(this);
    this.onUp = this.onUp.bind(this);
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  componentDidUpdate(prevProps: DraggableProps, prevState: DraggableState) {
    const { dragging, beingDragged } = this.state;
    if (dragging && !prevState.dragging) {
      document.addEventListener("mousemove", this.onMouseMove);
      document.addEventListener("mouseup", this.onUp);
      document.addEventListener("touchmove", this.onTouchMove, { passive: false });
      document.addEventListener("touchend", this.onUp);
    } else if (!dragging && prevState.dragging) {
      document.removeEventListener("mousemove", this.onMouseMove);
      document.removeEventListener("mouseup", this.onUp);
      document.removeEventListener("touchmove", this.onTouchMove);
      document.removeEventListener("touchend", this.onUp);
    }

    if (beingDragged) {
      document.addEventListener("click", this.handleOnClick, true);
    } else {
      document.removeEventListener("click", this.handleOnClick, true);
    }
  }

  delayedSetter() {
    setTimeout(
      (() => {
        this.setState({ beingDragged: false });
        document.removeEventListener("click", this.handleOnClick, true);
      })
      .bind(this),
      50
    );
  }

  handleOnClick(e) {
    e.stopPropagation();
    e.preventDefault();
  }

  onMouseDown(e: MouseEvent) {
    // only left mouse button
    if (e.button !== 0) return;
    const pos = this.myRef.getBoundingClientRect();
    this.onDragSelection(e);
    e.stopPropagation();
    e.preventDefault();
  }

  onTouchStart(e: TouchEvent) {
    // only when an item is touched
    if (e.changedTouches.length !== 1) return;
    const item = e.changedTouches[0];
    this.onDragSelection(item);
    e.stopPropagation();

    // Fix iOS issues where pages scrolls while dragging
    if (isIOS()) {
      e.preventDefault();
    }
  }

  onDragSelection(e: MouseEvent | TouchEventItem) {
    const pos = this.myRef.getBoundingClientRect();
    this.setState({
      dragging: true,
      rel: {
        x: e.pageX - pos.right,
        y: e.pageY - pos.bottom
      }
    });
  }

  onUp(e: MouseEvent | TouchEvent) {
    const { wasMoved } = this.state;
    this.setState({ dragging: false, wasMoved: false },
        () => this.delayedSetter()
      );
    e.stopPropagation();

    // Fix iOS issue where button is inaccessible after dragging
    if (!wasMoved && isIOS()) {
      e.target.dispatchEvent(new MouseEvent("click", {
        view: window,
        bubbles: true,
        cancelable: true
      }));
    }
  }

  onMouseMove(e: MouseEvent) {
    this.onMove(e);
    e.stopPropagation();
    e.preventDefault();
  }

  onTouchMove(e: TouchEvent) {
    // only when an item is touched
    if (e.changedTouches.length !== 1) return;
    const item = e.changedTouches[0];
    this.onMove(item);
    e.stopPropagation();
    e.preventDefault();
  }

  onMove(e: MouseEvent | TouchEventItem) {
    const { dragging, rel } = this.state;
    if (!dragging) return;
    const { updatePosition } = this.props;
    const x = Math.max(window.innerWidth - e.pageX + rel.x, 0);
    const y = Math.max(window.innerHeight - e.pageY + rel.y, 0);

    this.setState({
      pos: { x, y },
      beingDragged: true,
      wasMoved: true
    });

    if (updatePosition) {
      updatePosition(x, y);
    }
  }

  render() {
    const { pos } = this.state;
    const { children } = this.props;
    const divStyle = {
      position: "fixed",
      right: `${pos.x}px`,
      bottom: `${pos.y}px`
    };

    // Because Preact children is defined so oddly -_-
    const numChildren = Object.keys(children).length;
    const newChildren = [];
    let i: number;
    for (i = 0; i < numChildren; i = i + 1) {
      if (children[i]) {
        newChildren.push(cloneElement(children[i],
          { dragPosition: pos, ...outlineSuppressionHandlers }));
      }
    }

    return(
      <div
        onMouseDown={this.onMouseDown}
        onTouchStart={this.onTouchStart}
        style={divStyle}
        ref={c => this.myRef = c}
        role="presentation"
      >
        {newChildren}
      </div>
    );
  }
}

export const draggability = ({ isDraggable, updatePosition, x, y, children }) => (isDraggable ?
  (
    <Draggable updatePosition={updatePosition} x={x} y={y}>
      {children}
    </Draggable>
  )
  : <div>{children}</div>
);
