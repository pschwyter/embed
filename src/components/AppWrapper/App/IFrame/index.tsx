import { Component, h } from "preact";
import { capitalize } from "services/strings";
import postMessage from "services/postMessage";
import "./style.scss";

interface InterfaceIFrame {
  handle: string,
  iframeRef?: HTMLIFrameElement,
  chatURL: string,
  parentElement?: string | HTMLElement,
  styles?: string,
  metaFields?: object,
  isDrawerOpen: boolean,
  introShown: boolean,
  setIFrameRef(ref: HTMLIFrameElement): void,
  setIFrameLoaded(): void
}

export default class IFrame extends Component<InterfaceIFrame> {
  constructor(props: InterfaceIFrame) {
    super(props);

    this.onLoad = this.onLoad.bind(this);
    this.handleSetRef = this.handleSetRef.bind(this);
  }

  onLoad() {
    const {
      chatURL,
      parentElement,
      styles,
      iframeRef,
      setIFrameLoaded,
      introShown
    } = this.props;

    setIFrameLoaded();

    const toSend = {
      ...(styles ? { styles } : {}),
      ...(!parentElement ? { showCloseButton: true  } : {}),

      // Web interactions analytics data:
      // Indicates whether intro was shown to user
      introShown,
      // URL of the page where chat button was clicked
      initialURL: window.location.href
    };

    postMessage(iframeRef, toSend, chatURL);
  }

  handleSetRef(ref: HTMLIFrameElement) {
    const { setIFrameRef } = this.props;

    setIFrameRef(ref);
  }

  render() {
    const {
      chatURL,
      handle,
      isDrawerOpen
    } = this.props;

    return (
      <iframe
        name="ada-embed-iframe"
        className="ada-embed-iframe"
        src={chatURL}
        title={`${capitalize(handle)} Support Chat`}
        onLoad={this.onLoad}
        ref={this.handleSetRef}
        tabIndex={isDrawerOpen ? 0 : -1}
        allowFullScreen={true}
      />
    );
  }
}
