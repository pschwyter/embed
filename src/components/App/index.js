import { h, Component } from "preact";
import Drawer from "./Drawer";
import Button from "./Button";
import "./style.scss";

export default class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isDrawerOpen: false
    }
  }

  render(props) {
    const { handle } = props;

    return (
      <div>
        <Drawer handle={handle} />
        <Button />
      </div>
    );
  }
}
