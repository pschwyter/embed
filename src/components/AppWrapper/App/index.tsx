import Client from "models/Client";
import { Component, h, render } from "preact";
import checkRollout from "services/checkRollout";
import httpRequest from "services/httpRequest";
import postMessage from "services/postMessage";
import constructURL from "services/constructURL";
import { showZendeskWidget } from "services/zendesk";
import { draggability as Draggability } from "./Draggable";
import classnames from "classnames";
import Button from "./Button";
import IntroBlurb from "./IntroBlurb";
import Drawer from "./Drawer";
import IFrame from "./IFrame";
import {
  ADA_EVENT_BLUR,
  ADA_EVENT_FOCUS,
  ADA_EVENT_RESET,
  ADA_EVENT_TOGGLE,
  ADA_EVENT_GET_INFO,
  ADA_EVENT_GIVE_INFO,
  ADA_EVENT_DELETE_HISTORY,
  ADA_EVENT_SET_META_FIELDS
} from "constants/events";
import { InterfaceState as AppWrapperInterfaceState } from "components/AppWrapper";
import "./style.scss";
import { store, retrieve, removeStore } from "services/store";
import { CHATTER_TOKEN, CHATTER_CREATED, CHATTER_ZD_SESSION } from "constants/store";

interface UnreadMessage {
  amount: number
}

interface InterfaceApp extends AppWrapperInterfaceState {
  setAppState(hash: object, callback?: () => void): void
}

export default class App extends Component<InterfaceApp> {
  isInMobile: boolean;
  openChatInNewWindow: boolean;
  APIURL: string;
  connectorURL: string;
  adaModalElement: HTMLDivElement;
  appRef: HTMLDivElement;
  documentBodyOverflow: string;
  documentBodyPosition: string;
  documentBodyTop: string;
  documentBodyBottom: string;
  documentBodyLeft: string;
  documentBodyRight: string;
  pageYOffset: number;
  chatterToken: string;
  chatterCreated: string;
  chatterZDSession: string;

  constructor(props: InterfaceApp) {
    super(props);

    const {
      mobileOverlay,
      handle,
      cluster,
      domain
    } = props;

    this.toggleChat = this.toggleChat.bind(this);
    this.isInMobile = (
      navigator.userAgent.match(/(iPhone)|(iPod)|(android)|(webOS)/i) !== null
    );
    this.openChatInNewWindow = !mobileOverlay && this.isInMobile;

    this.setIFrameRef = this.setIFrameRef.bind(this);
    this.receiveMessage = this.receiveMessage.bind(this);
    this.handleAdaEvent = this.handleAdaEvent.bind(this);
    this.handleIntroShown = this.handleIntroShown.bind(this);
    this.setIFrameLoaded = this.setIFrameLoaded.bind(this);
    this.updateButtonPosition = this.updateButtonPosition.bind(this);
    this.lockDocumentBodyFromScrolling = this.lockDocumentBodyFromScrolling.bind(this);

    this.documentBodyOverflow = window.document.body.style.overflow;
    this.documentBodyPosition = window.document.body.style.position;
    this.documentBodyTop = window.document.body.style.top;
    this.documentBodyBottom = window.document.body.style.bottom;
    this.documentBodyLeft = window.document.body.style.left;
    this.documentBodyRight = window.document.body.style.right;

    const route = "connect";
    this.connectorURL = constructURL(
      { handle, cluster, route, domain },
      false
    );

    this.APIURL = constructURL(this.URLParams, true);
  }

  componentDidMount() {
    this.fetchClientAndSetup();
    this.initiateAdaEventListener();
  }

  componentWillUnmount() {
    this.removeListeners();
  }

  /**
   * Bind the message event listener to global scope
   */
  initiateMessageListener() {
    window.addEventListener("message", this.receiveMessage, false);
  }

  /**
   * Bind the ada event listener to global scope
   */
  initiateAdaEventListener() {
    document.addEventListener("ada-event", this.handleAdaEvent, false);
  }

  /**
   * Remove the event listeners to global scope
   */
  removeListeners() {
    document.removeEventListener("ada-event", this.handleAdaEvent, false);
    window.removeEventListener("message", this.receiveMessage, false);
  }

  /**
   * Handle incoming post message events from Chat
   */
  receiveMessage(event: MessageEvent) {
    const { client, chatURL } = this.props;

    // Ensure that event origin is the same as the Chat URL
    if (chatURL && !chatURL.startsWith(event.origin)) { return; }

    const {
      liveHandoff,
      zendeskLiveHandoff,
      chatter,
      analytics,
      closeChat,
      chatterIds,
      newMessages,
      created,
      zdSession
    } = event.data;

    const {
      isDrawerOpen,
      liveHandoffCallback,
      chatterTokenCallback,
      analyticsCallback
    } = this.props;

    if (liveHandoff && liveHandoffCallback) {
      liveHandoffCallback(liveHandoff);
    } else if (zendeskLiveHandoff) {
      showZendeskWidget(zendeskLiveHandoff, this.toggleChat);
    } else if (chatter) {
      store(client, CHATTER_TOKEN, chatter);
      this.chatterToken = chatter;

      if (chatterTokenCallback) {
        chatterTokenCallback(chatter);
      }
    } else if (analytics && analyticsCallback) {
      analyticsCallback(analytics);
    } else if (closeChat && isDrawerOpen) {
      this.toggleChat();
    } else if (chatterIds) {
      this.fetchChatterAndSetup(chatterIds);
    } else if (newMessages) {
      this.handleNewMessages(newMessages);
    } else if (zdSession) {
      store(client, CHATTER_ZD_SESSION, zdSession);
      this.chatterZDSession = zdSession;
    }

    if (created) {
      store(client, CHATTER_CREATED, created);
      this.chatterCreated = created;
    }
  }

  /**
   * Fetch chatter token from local/session storage via the connector's postMessage
   * and store the appropriate ID
   */
  fetchChatterAndSetup(chatterIds: string) {
    const {
      client
    } = this.props;
    const persistence = client.persistence === "normal" ? "local" : client.persistence;
    if (chatterIds[persistence]) {
      this.props.setAppState({
        chatter: chatterIds[persistence]
      }, () => this.fetchUnread());
    }
  }

  /**
   * Set the chatURL in state. We only want to update the ChatURL when Embed
   * initially loaded, or when reset. Otherwise, state changes can cause the Chat
   * iFrame to erroneously reload mid lifecycle.
   */
  setChatURL() {
    const {
      /** @type Client */
      client,
      setAppState
    } = this.props;

    const chatterToken = this.chatterToken;
    const chatterCreated = this.chatterCreated;
    const chatterZDSession = this.chatterZDSession;

    const chatURL = constructURL({
      ...this.URLParams,
      chatterToken,
      chatterCreated,
      chatterZDSession,
      followUpResponseId: client.intro && client.intro.response_id
    }, false);

    setAppState({
      chatURL
    });
  }

  get URLParams() {
    const {
      handle,
      domain,
      cluster,
      language,
      greeting,
      metaFields,
      introShown,
      privateMode,
      resetChatHistory
    } = this.props;

    return {
      handle,
      cluster,
      language,
      greeting,
      resetChatHistory,
      domain,
      privateMode,
      introShown,
      metaFields,
      initialURL: window.location.href
    };
  }

  /**
   * Fetch the client model and set state accordingly
   */
  fetchClientAndSetup() {
    const { handle, parentElement } = this.props;

    httpRequest({
      url: this.APIURL
    }).then((response) => {
      const { client } = response;

      this.props.setAppState({
        client: new Client(client)
      }, () => {
        const { rollout, chat: chatIsOn } = client;

        if (!chatIsOn) {
          console.warn("Sorry, please turn on the web chat integration in your bot's settings");
          return;
        }

        // Load Chatter info from storage
        this.chatterToken = retrieve(client, CHATTER_TOKEN);
        this.chatterCreated = retrieve(client, CHATTER_CREATED);
        this.chatterZDSession = retrieve(client, CHATTER_ZD_SESSION);

        // Need to set chatURL after getting client, but before setting shouldLoadEmbedUI
        this.setChatURL();

        // It should be loaded if rollout returns true, or if a parentElement is being used
        const shoudLoadEmbedUI = Boolean(parentElement) || checkRollout(rollout, handle);

        this.props.setAppState({
          shoudLoadEmbedUI
        }, () => {
          // AdaReadyCallback should only get triggered once Client has returned,
          // and once tokens have been retrieved.
          //
          // We need a setTimeout here to "pause" to allow rendering to catch up.
          // This was causing an issue where adaEmbed.reset() could not be called from
          // the adaReadyCallback.
          setTimeout(() => {
            this.triggerAdaReadyCallback();
          }, 0);

          if (shoudLoadEmbedUI) {
            this.initiateMessageListener();

            if (client.intro) {
              setTimeout(() => {
                this.props.setAppState({
                  showIntro: true
                });
              }, client.intro.delay * 1000);
            }
          }
        });
      });
    }, (error) => {
      console.warn(error);
      throw Error("An error occurred while retrieving the client");
    });
  }

  fetchUnread() {
    const route = `chatters/${this.chatterToken}/notification_status`;
    const url = constructURL(
      Object.assign(this.URLParams, { route }),
      true
    );
    httpRequest({
      url
    }).then((response) => {
      const { drawerHasBeenOpened } = this.props;
      // Set unread amount, connected to connector on chat
      // and open chat if in (active/pending) live state
      this.props.setAppState({
        unreadMessages: response.unread_amount,
        hasConnectedChat: true,
        drawerHasBeenOpened: drawerHasBeenOpened || response.is_live_state
      });
    }, (error) => {
      console.warn(error);
      throw Error("An error occurred while retrieving the client");
    });
  }

  clearChatterInfo() {
    const { client } = this.props;

    this.chatterToken = null;
    this.chatterCreated = null;
    this.chatterZDSession = null;
    // Remove Chatter token and created from storage
    removeStore(client, CHATTER_TOKEN);
    removeStore(client, CHATTER_CREATED);
    removeStore(client, CHATTER_ZD_SESSION);
  }

  /**
   * Event hanlder for custom Ada events
   */
  handleAdaEvent(event: CustomEvent) {
    const {
      chatURL,
      iframeRef,
      isDrawerOpen,
      parentElement,
      isIFrameLoaded,
      drawerHasBeenOpened,
      afterIFrameLoadsTasks
    } = this.props;
    const { detail } = event;
    const { type, data } = detail;
    const eventRequiresIFrame =
      [ADA_EVENT_SET_META_FIELDS, ADA_EVENT_DELETE_HISTORY].indexOf(type) > -1;

    if (eventRequiresIFrame) {
      if (!isIFrameLoaded) {
        afterIFrameLoadsTasks.push(event);

        this.props.setAppState({
          afterIFrameLoadsTasks
        });
      } else {
        switch (type) {
          case ADA_EVENT_SET_META_FIELDS:
            postMessage(iframeRef, data, chatURL);
            return;

          case ADA_EVENT_DELETE_HISTORY:
            // Remove the stored chatter info
            this.clearChatterInfo();
            postMessage(iframeRef, ADA_EVENT_DELETE_HISTORY, chatURL);
            return;
        }
      }
    } else {
      switch (type) {
        case ADA_EVENT_TOGGLE:
          this.toggleChat();
          return;

        case ADA_EVENT_RESET:
          const {
            resetChatHistory = true,
            metaFields = {},
            language = "",
            greeting = ""
          } = data || {};

          // Remove the stored chatter info
          this.clearChatterInfo();

          /**
           * In order to reset Chat we need to remove the IFrame component and re-render it.
           * To do this, we set `forceIFrameReRender` to `false`, then immediately back to `true`.
           * We can simulatenously set new values for language, greeting, and metaFields.
           */
          this.props.setAppState({
            language,
            greeting,
            metaFields,
            resetChatHistory,
            forceIFrameReRender: false
          }, () => {
            // Need to reset the chatURL before we re-open to ensure new query params are set
            this.setChatURL();

            this.props.setAppState({
              forceIFrameReRender: true
            });
          });
          return;

        case ADA_EVENT_GET_INFO:
          const outwardEvent = new CustomEvent(
            "ada-event-outward",
            {
              detail: {
                type: ADA_EVENT_GIVE_INFO,
                data: {
                  isDrawerOpen,
                  hasActiveChatter: Boolean(this.chatterToken),
                  hasClosedChat: drawerHasBeenOpened && !isDrawerOpen,
                  isChatOpen: isDrawerOpen || Boolean(isIFrameLoaded && parentElement)
                }
              },
              bubbles: true,
              cancelable: true
            }
          );
          this.appRef.dispatchEvent(outwardEvent);
          return;
      }
    }
  }

  /**
   * Handles incoming messages from Chat
   */
  handleNewMessages(messages: UnreadMessage) {
    const { unreadMessages } = this.props;

    this.props.setAppState({
      unreadMessages: unreadMessages + messages.amount
    });
  }

  /**
   * Handles clearing unread messages on both Embed and API
   */
  handleClearUnreadMessages() {
    this.props.setAppState({
      unreadMessages: 0
    });

    if (!this.chatterToken) return;
    // Send request to API to clear unread messages
    const route = `chatters/${this.chatterToken}/live_chat_unread_amount`;
    const url = constructURL(
      Object.assign(this.URLParams, { route }),
      true
    );
    httpRequest({
      url,
      method: "DELETE"
    });
  }

  updateButtonPosition(x: number, y: number) {
    this.props.setAppState({
      buttonPosition: { x, y }
    });
  }

  /**
   */
  handleIntroShown() {
    this.props.setAppState({
      introShown: true
    });
  }

  /**
   * Lock the document body from scrolling. If we don't do this,
   * there are SERIOUS issues on iOS.
   */
  lockDocumentBodyFromScrolling(event) {
    const { isDrawerOpen } = this.props;
    const nextIsDrawerOpen = !isDrawerOpen;

    if (!this.isInMobile || nextIsDrawerOpen || event.propertyName !== "transform") {
      return;
    }

    // save current page position so we can scroll back there
    this.pageYOffset = window.pageYOffset;

    window.document.body.style.overflow = "hidden";
    window.document.body.style.position = "fixed";
    window.document.body.style.top = "0";
    window.document.body.style.bottom = "0";
    window.document.body.style.left = "0";
    window.document.body.style.right = "0";
  }

  /**
   * Set back intial values from client document body
   */
  unlockDocumentBodyFromScrolling() {
    window.document.body.style.overflow = this.documentBodyOverflow;
    window.document.body.style.position = this.documentBodyPosition;
    window.document.body.style.top = this.documentBodyTop;
    window.document.body.style.bottom = this.documentBodyBottom;
    window.document.body.style.left = this.documentBodyLeft;
    window.document.body.style.right = this.documentBodyRight;

    // scroll the page back to where it was before chat opened
    window.scrollTo(0, this.pageYOffset);
  }

  /**
   * Open/close the Drawer component, or open a new window if in mobile
   */
  toggleChat() {
    const { isDrawerOpen, iframeRef, chatURL } = this.props;
    const nextIsDrawerOpen = !isDrawerOpen;

    // Clear unread messages
    this.handleClearUnreadMessages();

    // Open Chat in a new window if mobile
    if (this.openChatInNewWindow) {
      window.open(chatURL);

      return;
    }

    // Unlock body from scrolling on mobile
    // locking happens in an event listener in the Drawer component
    if (this.isInMobile) {
      if (!nextIsDrawerOpen) {
        // Unlock body from scrolling
        this.unlockDocumentBodyFromScrolling();
      }
    }

    this.props.setAppState({
      isDrawerOpen: nextIsDrawerOpen,
      drawerHasBeenOpened: true
    }, () => {
      if (iframeRef) {
        if (nextIsDrawerOpen) {
          // To ensure the input bar is always focused in Chat when the drawer is opened
          iframeRef.contentWindow.focus();
          postMessage(iframeRef, ADA_EVENT_FOCUS, chatURL);
        } else {
          postMessage(iframeRef, ADA_EVENT_BLUR, chatURL);

          if (document.activeElement instanceof HTMLElement) {
            document.activeElement.blur();
          }
        }
      }
    });
  }

  /**
   * Set the iFrame ref on the parent
   */
  setIFrameRef(ref: HTMLIFrameElement) {
    this.props.setAppState({
      iframeRef: ref
    });
  }

  setIFrameLoaded() {
    const { afterIFrameLoadsTasks } = this.props;

    this.props.setAppState({
      isIFrameLoaded: true,
      afterIFrameLoadsTasks: []
    }, () => {
      afterIFrameLoadsTasks.forEach((event) => {
        this.handleAdaEvent(event);
      });
    });
  }

  triggerAdaReadyCallback() {
    const { adaReadyCallback } = this.props;
    if (adaReadyCallback) { adaReadyCallback(); }
  }

  /**
   * To be rendered when user is using parentElement config
   */
  renderIFrameForParentElement() {
    const {
      iframeRef,
      isDrawerOpen,
      introShown
    } = this.props;

    return (
      <IFrame
        {...this.props}
        iframeRef={iframeRef}
        isDrawerOpen={isDrawerOpen}
        setIFrameRef={this.setIFrameRef}
        setIFrameLoaded={this.setIFrameLoaded}
        introShown={introShown}
      />
    );
  }

  /**
   * Standard config render (no parentElement)
   */
  renderStandardConfigElements() {
    const { mobileOverlay, hideMask, dragAndDrop } = this.props;
    const {
      client,
      showIntro,
      iframeRef,
      introShown,
      isDrawerOpen,
      buttonPosition,
      unreadMessages,
      hasConnectedChat,
      drawerHasBeenOpened
    } = this.props;

    return (
      <div>
        {/* Do not render Drawer if Chat will be opened in new window */}
        {!this.openChatInNewWindow && (
          <Drawer
            {...this.props}
            hideMask={hideMask}
            iframeRef={iframeRef}
            introShown={introShown}
            isDrawerOpen={isDrawerOpen}
            toggleChat={this.toggleChat}
            setIFrameRef={this.setIFrameRef}
            setIFrameLoaded={this.setIFrameLoaded}
            drawerHasBeenOpened={drawerHasBeenOpened}
            useMobileOverlay={mobileOverlay && this.isInMobile}
            transitionEndHandler={this.lockDocumentBodyFromScrolling}
          />
        )}
        <Draggability
          x={buttonPosition.x}
          y={buttonPosition.y}
          updatePosition={this.updateButtonPosition}
          isDraggable={dragAndDrop}
        >
          {showIntro && client.intro.style.toLowerCase() === "text" && !drawerHasBeenOpened && (
            <IntroBlurb
              client={client}
              toggleChat={this.toggleChat}
              isInMobile={this.isInMobile}
              isDraggable={dragAndDrop}
              onShow={this.handleIntroShown}
            />
          )}
          {(!isDrawerOpen || dragAndDrop) && (
            <Button
              client={client}
              toggleChat={this.toggleChat}
              showIntroEmoji={
                showIntro &&
                client.intro.style.toLowerCase() === "emoji" &&
                !drawerHasBeenOpened
              }
              showNotification={unreadMessages > 0}
              isDraggable={dragAndDrop}
            />
          )}
        </Draggability>
        {/* This iFrame is here to connect with /chat/connect/ and pull the chatter's ID from
            both local and session storage through postMessage. Once the message is received
            hasConnectedChat will be set to true and the iFrame will no longer be rendered.
            For more info @Kiwi or @Nic*/}
        {!hasConnectedChat && (
          <iframe
            name="ada-embed-connector-iframe"
            className="ada-embed-connector-iframe"
            src={this.connectorURL}
            title="Ada Embed Connector"
            style="display: none;"
          />
        )}
      </div>
    );
  }

  get elementToRender() {
    const { parentElement } = this.props;
    const {
      shoudLoadEmbedUI,
      forceIFrameReRender
    } = this.props;

    if (!shoudLoadEmbedUI || !forceIFrameReRender) {
      return null;
    }

    if (parentElement) {
      return this.renderIFrameForParentElement();
    }

    return this.renderStandardConfigElements();
  }

  render() {
    const { parentElement } = this.props;

    return (
      <div
        id="ada-embed"
        className={classnames("ada-embed-app",
          {
            "ada-embed-app--inside-parent": parentElement
          }
        )}
        ref={elem => this.appRef = elem}
      >
        {this.elementToRender}
      </div>
    );
  }
}
