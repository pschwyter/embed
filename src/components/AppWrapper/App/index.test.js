/* eslint-env node, jest */

import "jsdom-global/register";
import { shallow } from "preact-render-spy";
import App from "./index";
import IFrame from "./IFrame";

function setup(propsOverride = {}) {
  const defaultsProps = {
    handle: "nic",
    setAppState: jest.fn()
  };

  const props = Object.assign(defaultsProps, propsOverride);
  const PRSWrapper = shallow(<App {...props} />);

  return {
    props,
    PRSWrapper
  };
}

jest.mock("services/httpRequest");

describe("<App />", () => {
  it("should match snapshot", done => {
    const { PRSWrapper } = setup();

    // We need to wrap in setTimeout and use 'done' because setState is operated asynchronously with preact-render-spy
    setTimeout(() => {
      expect(PRSWrapper).toMatchSnapshot();
      done();
    }, 0);
  });

  it("should no longer render connector iframe if hasConnectedChat is true", done => {
    const { PRSWrapper } = setup({
      hasConnectedChat: true
    });

    setTimeout(() => {
      expect(PRSWrapper.find(".ada-connector-iframe").length).toBe(0);
      done();
    }, 0);
  });

  it("should position the iFrame inside parentElement if specified", done => {
    let dummyElement = document.createElement("div");

    const { PRSWrapper } = setup({
      parentElement: dummyElement,
      shoudLoadEmbedUI: true,
      forceIFrameReRender: true
    });

    setTimeout(() => {
      expect(PRSWrapper.find(<IFrame />).length).toBeTruthy();
      done();
    }, 0);
  });

  it("should apply a class of 'ada-embed-app--inside-parent' when a parentElement is specified", done => {
    const { PRSWrapper } = setup({
      parentElement: "parent-element"
    });

    setTimeout(() => {
      expect(PRSWrapper.find(".ada-embed-app--inside-parent").length).toBe(1);
      done();
    }, 0);
  });

  it("should not have the class 'ada-embed-app--inside-parent' when no parentElement specified", done => {
    const { PRSWrapper } = setup();

    setTimeout(() => {
      expect(PRSWrapper.find(".ada-embed-app--inside-parent").length).toBe(0);
      done();
    }, 0);
  });
});
