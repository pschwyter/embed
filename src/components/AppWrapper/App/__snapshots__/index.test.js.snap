// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<div
  id="ada-embed"
  class="ada-embed-app"
>
</div>

`;
